PEP 704: Require virtual environments by default for installers (#2964)
python/peps: pep-0704.rst

PEP: 704

> Title: Require virtual environments by default for installers

Reviewer: ```suggestion
Title: Require virtual environments by default for package installers
```

@ned-deily suggested this to avoid any confusion with the installers for Python itself, and I agree FWIW

Author: Pradyun Gedam <pradyunsg@gmail.com>

> Sponsor: Brett Cannon <brett@python.org>

Reviewer 2: Yep, I agree.

Motivation
==========

Python virtual environments are an essential part of the development workflow for Python. However, it requires extra effort since it is an opt-in feature, and requires users to either:

- take explicit steps to activate/deactivate a virtual environment
- use ``<path-to-venv>/<bin-path>/<executable>`` to run files

This creates a barrier to entry for many new users, since virtual environments serve as a pre-requisite for typical Python development workflows and how virtual environments work is a lot of information for anyone new. It can take a lot of extra time and effort to explain them.

Further, activation of virtual environment uses a slightly different syntax/mechanism to activate on different platforms.

It also creates a scope for mistakes, since users must remember to activate the virtual environment before running an installer like ``pip``. On certain Linux distributions, this can result in the installer modifying files that are owned by the operating system (this is partially mitigated by :pep:`668` for distributions that opt-in to marking their environments accordingly).

> This PEP proposes that the default behaviour of installers like ``pip`` should be to require a virtual environment to be active. This would make it easier for new users to get started with Python, and would reduce the scope for mistakes.

Reviewer: These two sentences are really Rationale rather than Motivation, as they try to explain why the solution you're proposing should be accepted to solve the problem outlined in the Motivation. Furthermore, I'd strongly suggest expanding that Rationale section at least somewhat so it actually explains how implementing this default behavior will achieve the aim you state here, as currently on the surface it appears to contradict itself (which, to be clear, I believe you have a reasonably plausible case for—you just have to actually make it :slightly_smiling_face: ).

In particular, most of the aforementioned Motivation focuses at some length on the UX challenges that virtual environment present for new users, as it describes the extra steps that using venvs require, states that this "creates a barrier to entry for many new users", and mentions that "how virtual environments work is a lot of information for anyone new. It can take a lot of extra time and effort to explain them."

However, in the Rationale section, it states having installers "require a virtual environment to be active" by default "would make it easier for new users to get started with Python". As the previous lines just established that understanding venvs and knowing the actions to take to create and active one are "a barrier to entry for many new users", and there is no mention here of _how_ this would be made easier for new users, it sounds like a _prima facie_ contradiction that requiring new users to create and activate a venv before installing any package would make it easier, rather than harder, "for new users to get started".

Also, by contrast, the Motivation more briefly touches on the issue of accidentally installing packages without a venv active, mentioning a single concrete example that is at least partially mitigated by another accepted PEP, despite this arguably being the broader and more obvious benefit here.

Therefore, I suggest:

* Make a Rationale section starting with this paragraph.
* Mention how this proposal will actually make thins easier for new users by, e.g.,
  * Mandating that tools give users simple, specific and actionable advice on how to create and activate one, and links to help resources
  * Standardizes the naming of venvs to avoid user confusion and improve tool automation
  * Etc.
* Add at least >=one more example to the Motivation of the real-world problems not activating a venv causes users

Specification
=============

> When a user runs an installer without an active virtual environment, the installer MUST print an error message and exit with a non-zero exit code. The error message MUST inform the user that a virtual environment is required, SHOULD provide shell-specific instructions on how to create and activate a virtual environment named ``.venv``, and SHOULD provide a link to a documentation page that explains how to create and activate a virtual environment.

Reviewer: What counts as a "virtual environment"? Only an environment created with `venv`/`virtualenv`? What about Conda environments? Or PEP 582 environments? Or other types of isolated environments? The PEP should either explicitly define this, or link (`:term:`) to a precise and authoritative definition, e.g. in the PyPA glossary.

> When a user runs an installer without an active virtual environment, the installer MUST print an error message and exit with a non-zero exit code. The error message MUST inform the user that a virtual environment is required, SHOULD provide shell-specific instructions on how to create and activate a virtual environment named ``.venv``, and SHOULD provide a link to a documentation page that explains how to create and activate a virtual environment.

Reviewer: Also, what counts as "an installer"? Obviously, `pip` counts (as its mentioned by name), and I'm _assuming_ Hatch, PDM and Poetry count as well when used in that capacity, while `apt`, `dnf`, `brew` and `choco` don't, though that isn't explicitly stated.

If you limit it to "Python-specific" installer, what about tools like shiv, pex, etc.? And what about, of course, Conda? The PEP should provide a precise definition of that term and examples of what tools would and would not quality (similar to PEP 668 for what qualifies as an externally-managed environment).

The installer SHOULD also provide a mechanism to disable this behaviour, allowing a user to opt-in to running the tool outside of a virtual environment. The installer MAY include this in the error message that it prints, and MUST include it in the documentation page it links to.

If the installer includes a link to a documentation page in the error message described above, the documentation page SHOULD be written with a focus on new users. It should explain what virtual environments are, why they are required, and how to create and activate a virtual environment. It SHOULD include instructions for the most common shells and platforms.

> Installers MAY choose to implement this default behaviour on any Python versions, but MUST implement it on Python 3.13 or newer.

Reviewer: IMO, it would be best to be explicit here that you mean the Python environment the installer is being instructed to install into (as I'm _assuming_ you mean, vs. e.g. the runtime environment of the installer, or the default/system Python.

Backwards Compatibility
=======================

> This PEP is backwards incompatible with workflows where users who are using installers outside of virtual environments. Such users will be prompted with an error message and will need to either:

Reviewer 3: ```suggestion
This PEP is backwards incompatible with workflows where users are using installers outside of virtual environments. Such users will be prompted with an error message and will need to either:
```

- explicitly opt-in to running the installer outside of a virtual environment, or
- create and use a virtual environment

Users who are already using virtual environments will not be affected by this change.

> Workflow tools (which manage virtual environments for the user, under the hood) should be unaffected, since they should already be using the virtual environment for running the installer.

Reviewer: They will be affected if they don't use `venv`/`virtualenv` virtual environments (e.g. Conda envs, I'd assume PDM PEP 582 envs, and possibly others), assuming tools only detect them. Either way, that really should be explicitly addressed here.

Reviewer: (But this could, of course, be deferred to a followup at your discretion, as it is really more of a content issue rather than editorial matter).

Security Implications
=====================

> This PEP does not introduce any new security implications.

Reviewer 4: I'm pretty sure it can be argued that this improves the security situation, as it will guide users to avoid to "just add `sudo`" to get things to work, and potentially installing arbitrary code with those privileges. By defaulting to environments, the user doesn't need an escape hatch for not having enough privileges to install something - things will work fine just with standard rights.

Reviewer 5: We already have a non-root option available, and it's the default in current versions of pip (user installation). This isn't really a security concern IMO.

> This PEP does not introduce any new security implications.

Reviewer: You could potentially argue that it could improve security slightly, as any malicious package will (at least initially) only be installed into a single isolated virtualenv, as opposed to the system environment used for all projects (perhaps even with administrator rights).

Reviewer 6: There's roughly no security win, since the ability to execute arbitrary code means you more or less have all the powers the user does.

Reviewer 4: > There's roughly no security win, since the ability to execute arbitrary code means you more or less have all the powers the user does.

Yeah, but it _does_ avoid the "admin escape hatch" that several users have learned to take if they cannot install something with their regular user, and it turns out they're trying to install to the system python. It's great if this is not done by default anymore, but it still helps even if you're accidentally installing something as root user to not mess up your system python, but put them in a separate env (though when you're installing something as root, obviously any potential security improvements aren't relevant anymore).

Author: > Yeah, but it does avoid the "admin escape hatch" that several users have learned to take if they cannot install something with their regular user, and it turns out they're trying to install to the system python. 

We've closed that with EXTERNALLY-MANAGED (PEP 668).

How to Teach This
=================

> This PEP does require that new users create and use a virtual environment to get started with using Python packages. This is however a best practice, as demonstrated by the section on "basics of how to install Python packages" in the Python Packaging User Guide, which explains how/what virtual environments are before discussing using ``pip``.[1]_

Reviewer: ```suggestion
This PEP does require that new users create and use a virtual environment to get started with using Python packages. This is, however, a best practice, as demonstrated by the `section on the basics of how to install Python packages <https://packaging.python.org/en/latest/tutorials/installing-packages/#creating-virtual-environments>`__ in the Python Packaging User Guide, which explains what virtual environments are before discussing using ``pip``.
```

* Fixed a couple grammar errors.

* The footnote syntax is broken.

* Further, It doesn't really make sense to require to user to find and click through to a footnote, just to have to click on another (bare) link. Instead, just link the content directly. Per PEP 12:

  > Native reST hyperlink targets or inline hyperlinks in the text should be used in preference to footnotes for including URLs to online resources.

  See also #2130

* In fact, you could normally just do `` :ref:`sections-ref-target-label-on-pypug` `` to link it directly, but unfortunately it doesn't (yet) have one.

Reference Implementation
========================

> There is no reference implementation for this PEP, however the proposed behaviour is already implemented in ``pip`` and can be achieved by setting ``PIP_REQUIRE_VENV`` environment variable to ``1`` (not setting it would result in the proposed opt-in behaviour).

Reviewer: ```suggestion
There is no reference implementation for this PEP. However, the proposed behaviour is already implemented in ``pip`` and can be activated by setting the ``PIP_REQUIRE_VENV`` environment variable to ``1``. (Leaving it unset results in the proposed opt-in behaviour of not requiring a virtual environment for installation.)
```

* Clarify what you mean by the last clause (I was very confused by it, and it took several re-reads to parse out)
* Fix several grammar and punctuation errors and issues
* Fix apparent typo ("achieved" -> "activated")

Reviewer: Looks like there were all missed when this suggestion was outdated by another change; I meant to update and re-make it, but didn't end up getting to it in time, sorry.

Rejected Ideas
==============

Do not specify a name for the virtual environment directory
-----------------------------------------------------------

Using a consistent name for the virtual environment directory is important for a few reasons:

> 1. It makes it easier for users to find the virtual environment directory, and to activate it.
> 2. It removes a decision point for new users, since they do not need to decide on a name for the virtual environment directory.
> 3. It creates a clear convention within the ecosystem, which makes it easier for users to find documentation.
> 4. It ensures consistency across different tools, so that differences in the error message of different tools do not confuse users.

Reviewer: ```suggestion
* It makes it easier for users to find the virtual environment directory, and to activate it.
* It removes a decision point for new users, since they do not need to decide on a name for the virtual environment directory.
* It creates a clear convention within the ecosystem, which makes it easier for users to find documentation.
* It ensures consistency across different tools, so that differences in the error messages do not confuse users.
```

* Use unordered lists for items that don't have a clear and unambiguous sequential order (which also avoids it inevitably getting out of sync if items are added, moved or removed)
* Fix grammar error and avoid repetition

Author: I prefer the current style.

Use a different name for the virtual environment directory
----------------------------------------------------------

Functionally, the directory name does not matter much as long as there is a single consistent suggestion.

The name ``.venv`` was picked since it:

> 1. does not conflict with any valid Python import name
> 2. does not conflict ``venv`` module in the standard library
> 3. has pre-existing usage in the Python community
> 4. has support for auto-detection in common text editors
> 5. can be typed without modifier keys on common keyboard layouts

Reviewer: ```suggestion
* does not conflict with any valid Python import name
* does not conflict ``venv`` module in the standard library
* has pre-existing usage in the Python community
* has support for auto-detection in common text editors
* can be typed without modifier keys on common keyboard layouts
```

Same here

Author: I prefer the numbered list here, since this is an ordered list to me.

Reviewer: Fair enough

Do not couple tooling behaviour with a Python version
-----------------------------------------------------

This PEP creates a coupling between the behaviour of installers and the Python version, enforcing a behaviour change in installers based on a Python version.

> This is already a rollout mechanism being used for behaviour changes in the installation tooling. For example, ``pip`` on Python 3.11 will use ``importlib.metadata`` instead of ``pkg_resources`` for parsing/fetching package metadata, and ``sysconfig`` instead of ``distutils.sysconfig`` for getting the paths to unpack wheels into.

Reviewer: ```suggestion
This is already a rollout mechanism being used for behaviour changes in the installation tooling. For example, ``pip`` on Python 3.11 will use :mod:`importlib.metadata` instead of :mod:`!pkg_resources` for fetching/parsing package metadata, and :mod:`sysconfig` instead of :mod:`!distutils.sysconfig` for getting the paths to unpack wheels into.
```

* Use appropriate markup/cross references
* Flip words ("fetching" -> "parsing") to logical order (fetching comes before parsing, no?)

Author: I don't think the markup references are particularly valuable -- is that a blocking concern?

It's primarily used for parsing rather than fetching; hence the order.

Reviewer: Okay, up to you :)

The difference with those cases is that they're supposed to be largely transparent to end users. This PEP is proposing a behaviour change that is not transparent to end users, and requires them to take action.

The primary benefit of this is that it allows for redistributors to adapt their tooling in time for the new Python version and provides a clear and consistent point for change across the ecosystem. It also puts a clear deadline on when the default behaviour will consistently require a virtual environment by default (once Python 3.12 goes end-of-life).

> The primary issue with this approach is that it enforces a behaviour change on users when they upgrade to a new Python version, which can hamper the adoption of a new Python version. However, this is a migration/upgrade for existing users and it is a common expectation that _some_ changes to be needed for migration/upgrades.

Reviewer: ```suggestion
The primary issue with this approach is that it enforces a behaviour change on users when they upgrade to a new Python version, which can hamper the adoption of a new Python version. However, this is a migration/upgrade for existing users and it is a common expectation that *some* changes will be needed for migration/upgrades.
```

* Fix syntax mistake
* Fix grammar error

The author of this PEP believes that enforcing the benefits of applying this consistently throughout the ecosystem with a deadline, outweigh the drawbacks of enforcing a best-practice on users when they upgrade.

Open Issues
===========

> None.
> 
> 
> Footnotes
> =========
> 
> .. [1] https://packaging.python.org/en/latest/tutorials/installing-packages/#creating-virtual-environments
> 
> Copyright

Reviewer: ```suggestion
None.

Copyright
```

Turned into a proper link; see above.

PEP: 704
Title: Require virtual environments by default for package installers
Author: Pradyun Gedam <pradyunsg@gmail.com>
Sponsor: Brett Cannon <brett@python.org>
PEP-Delegate: Paul Moore <p.f.moore@gmail.com>
Discussions-To: https://discuss.python.org/t/22846
Status: Draft
Type: Standards Track
Topic: Packaging
Content-Type: text/x-rst
Created: 16-Jan-2023
Post-History: `16-Jan-2023 <https://discuss.python.org/t/22846>`__

Abstract
========

This PEP recommends that package installers like ``pip``  require a virtual environment by default, on Python 3.13+.

Motivation
==========

Python virtual environments are an essential part of the development workflow for Python. However, they require extra effort since they are an opt-in feature, and requires users to either:

- take explicit steps to activate/deactivate a virtual environment
- use :samp:`{<path-to-venv>}/{<bin-path>}/{<executable>}` to run files

> For new users, things will seemingly work correctly when virtual environments are not used until they don't. Further, activating a virtual environment uses slightly different syntax and mechanisms on different platforms. This complicates the introduction to virtual environments since now information and context about how/why they are useful needs to explained to justify adding additional steps to the workflow.

Author: ```suggestion
For new users, things will seemingly work correctly when virtual environments are not used -— until they don't. Further, activating a virtual environment uses slightly different syntax and mechanisms on different platforms. This complicates the introduction to virtual environments, since now information and context about how/why they are useful needs to explained to justify adding additional steps to the workflow.
```

> It also creates a scope for mistakes, since users need to remember to activate the virtual environment before running an installer like ``pip`` or configure those installers to error out. On certain Linux distributions, forgetting to do so can result in the installer modifying files that are owned by the operating system (which is partially mitigated by :pep:`668` for distributions that opt-in to marking their environments accordingly).

Author: This feels wrong to me as a reader; so skipping it. :)

1. It makes it easier for users to find the virtual environment directory, and to activate it.
2. It removes a decision point for new users, since they do not need to decide on a name for the virtual environment directory.
3. It creates a clear convention within the ecosystem, which makes it easier for users to find documentation.
4. It ensures consistency across different tools, so that differences in the error message of different tools do not confuse users.

Use a different name for the virtual environment directory
----------------------------------------------------------

Functionally, the directory name does not matter much as long as there is a single consistent suggestion.

The name ``.venv`` was picked since it:

1. does not conflict with any valid Python import name
2. does not conflict ``venv`` module in the standard library
3. has pre-existing usage in the Python community
4. has support for auto-detection in common text editors
5. can be typed without modifier keys on common keyboard layouts

Do not couple tooling behaviour with a Python version
-----------------------------------------------------

> This PEP creates a coupling between the behaviour of installers and the Python version, enforcing a behaviour change in installers based on a Python version.

Reviewer: This repeats an nearly the same phrase twice (the first of which _also_ repeats the section title on the line above), which is seems thoroughly redundant. I therefore suggest either one or the other:

```suggestion
This PEP creates a coupling between the behaviour of installers and the Python version.
```

or 

```suggestion
This PEP enforces a behaviour change in installers based on a Python version.
```

This is already a rollout mechanism being used for behaviour changes in the installation tooling. For example, ``pip`` on Python 3.11 will use ``importlib.metadata`` instead of ``pkg_resources`` for parsing/fetching package metadata, and ``sysconfig`` instead of ``distutils.sysconfig`` for getting the paths to unpack wheels into.

The difference with those cases is that they're supposed to be largely transparent to end users. This PEP is proposing a behaviour change that is not transparent to end users, and requires them to take action.

The primary benefit of this is that it allows for redistributors to adapt their tooling in time for the new Python version and provides a clear and consistent point for change across the ecosystem. It also puts a clear deadline on when the default behaviour will consistently require a virtual environment by default (once Python 3.12 goes end-of-life).

> The primary issue with this approach is that it enforces a behaviour change on users when they upgrade to a new Python version, which can hamper the adoption of a new Python version. However, this is a migration/upgrade for existing users and it is a common expectation that *some* changes will be needed for migration/upgrades.

Reviewer: ```suggestion
The primary issue with this approach is that it enforces a behaviour change on users when they upgrade to a new Python version, which can hamper the adoption of a new Python version. However, this is a migration/upgrade for existing users, and it is a common expectation that *some* changes will be needed for migration/upgrades.
```

Somehow missed this missing comma before, sorry

Author: This feels wrong to me as a reader; so skipping it. :)

Reviewer: As above—it sounds wrong to me as a technical writer and copyeditor, and the same grammar applies—these are two independent clauses separated by a coordinating conjunction, so they need a comma. Particularly so since the second clause repeats a phrase ("migration/upgrades") from the previous clause, and the distance created by the comma also helps make this repetition feel less jarring and more intentional.

Also, fixed a plural/singular agreement error.

That being said, same as above applies—it's your PEP :)

Author: Let's leave this as-is.

> The author of this PEP believes that enforcing the benefits of applying this consistently throughout the ecosystem with a deadline, outweigh the drawbacks of enforcing a best-practice on users when they upgrade.

Reviewer: ```suggestion
The author of this PEP believes that the benefits of applying this consistently throughout the ecosystem with a deadline outweigh the drawbacks of enforcing a best-practice on users when they upgrade.
```

Remove spurious word and comma
